src/main/scala/mesosphere/marathon/storage/migration/MigrationTo200.scala

```diff
@@ -0,0 +1,151 @@
+      (__ \ "healthy").readNullable[Boolean]
+    ) { (condition, since, activeSince, healthy) =>
+        InstanceState(condition, since, activeSince, healthy)
+      }
+  }
+
+  /**
+    * Read format for old instance without goal.
+    */
+  val instanceJsonReads160: Reads[Instance] = {
+    (
+      (__ \ "instanceId").read[Instance.Id] ~
+      (__ \ "agentInfo").read[AgentInfo] ~
+      (__ \ "tasksMap").read[Map[Task.Id, Task]] ~
+      (__ \ "runSpecVersion").read[Timestamp] ~
+      (__ \ "state").read[InstanceState](instanceStateReads160) ~
+      (__ \ "unreachableStrategy").readNullable[raml.UnreachableStrategy] ~
+      (__ \ "reservation").readNullable[Reservation]
+    ) { (instanceId, agentInfo, tasksMap, runSpecVersion, state, maybeUnreachableStrategy, reservation) =>
+        val unreachableStrategy = maybeUnreachableStrategy.
+          map(Raml.fromRaml(_)).getOrElse(UnreachableStrategy.default())
+        new Instance(instanceId, agentInfo, state, tasksMap, runSpecVersion, unreachableStrategy, reservation)
+      }
+  }
+
+  implicit val instanceResolver: IdResolver[Instance.Id, JsValue, String, ZkId] =
+    new IdResolver[Instance.Id, JsValue, String, ZkId] {
+      override def toStorageId(id: Id, version: Option[OffsetDateTime]): ZkId =
+        ZkId(category, id.idString, version)
+
+      override val category: String = "instance"
+
+      override def fromStorageId(key: ZkId): Id = Instance.Id.fromIdString(key.id)
+
+      override val hasVersions: Boolean = false
+
+      override def version(v: JsValue): OffsetDateTime = OffsetDateTime.MIN
+    }
+
+  implicit val instanceJsonUnmarshaller: Unmarshaller[ZkSerialized, JsValue] =
+    Unmarshaller.strict {
+      case ZkSerialized(byteString) =>
+        Json.parse(byteString.utf8String)
+    }
+
+  /**
+    * This function traverses all instances in ZK and sets the instance goal field.
+    */
+  def migrateInstanceGoals(instanceRepository: InstanceRepository, persistenceStore: PersistenceStore[_, _, _])(implicit mat: Materializer): Future[Done] = {
+
+    logger.info("Starting goal migration to Storage version 200")
+
+    maybeStore(persistenceStore).map { store =>
+      instanceRepository
+        .ids()
+        .mapAsync(1) { instanceId =>
+          store.get[Instance.Id, JsValue](instanceId)
+        }
+        .via(migrationFlow)
+        .mapAsync(1) { updatedInstance =>
+          instanceRepository.store(updatedInstance)
+        }
+        .runWith(Sink.ignore)
+    } getOrElse {
+      Future.successful(Done)
+    }
+  }
+
+  /**
+    * Update the goal of the instance.
+    * @param instance The old instance.
+    * @return An instance with an updated goal.
+    */
+  def updateGoal(instance: Instance): Instance = {
+    val updatedInstanceState = if (!instance.hasReservation) {
+      instance.state.copy(goal = Goal.Running)
+    } else {
+      if (instance.isReservedTerminal) {
+        instance.state.copy(goal = Goal.Stopped)
```

Comment:
I'd like to understand this better. A `reservedTerminal` instance is one with condition `Reserved` and a task in a terminal state, correct? We might want to launch a new task for this instance, which previously was implicit, and bound to the instance count specified in the run spec.

Is the idea here that we'll migrate to `Stopped`, and recovery logic will eventually set this to `Running` after failover, once it realizes we're under capacity?

Reply:
If I understood you correctly, yes. `isReservedTerminal` means the tasks are not running but do have a reservation. The can be started again the goal will be `Running` again. See https://github.com/mesosphere/marathon/pull/6328/files#diff-4d03c99a6694604a192ac4812b53a108R126.

Reply:
Ok, so the `Stopped` doesn't really have any effect atm. Got it.